pkg/comp-functions/functions/vshnmariadb/user_management.go

```diff
@@ -0,0 +1,266 @@
+	}
+	if connectionDetailRef != nil {
+		om.Name = connectionDetailRef.Name
+		om.Namespace = connectionDetailRef.Namespace
+	}
+
+	userpassSecret := &corev1.Secret{
+		ObjectMeta: om,
+		Type:       corev1.SecretType("connection.crossplane.io/v1alpha1"),
+		Data: map[string][]byte{
+			"MARIADB_USERNAME": []byte(username),
+			"MARIADB_PASSWORD": userpassCD["userpass"],
+			"MARIADB_DB":       []byte(dbname),
+			"MARIADB_HOST":     compositeCD["MARIADB_HOST"],
+			"MARIADB_PORT":     compositeCD["MARIADB_PORT"],
+			"MARIADB_URL":      []byte(url),
+			"ca.crt":           compositeCD["ca.crt"],
+		},
+	}
+
+	err = svc.SetDesiredKubeObject(userpassSecret, fmt.Sprintf("%s-user-%s", comp.GetName(), username))
+	if err != nil {
+		svc.Log.Error(err, "cannot get userpassword from secret")
+		svc.AddResult(runtime.NewWarningResult(fmt.Sprintf("cannot get userpassword from secret: %s", err)))
+	}
+}
+
+func addProviderConfig(comp *vshnv1.VSHNMariaDB, svc *runtime.ServiceRuntime) {
+	cd := svc.GetConnectionDetails()
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "provider-conf-credentials",
+			Namespace: comp.GetInstanceNamespace(),
+		},
+		Data: map[string][]byte{
+			"username": cd["MARIADB_USERNAME"],
+			"password": cd["MARIADB_PASSWORD"],
+			"endpoint": cd["MARIADB_HOST"],
+			"port":     cd["MARIADB_PORT"],
+		},
+	}
+
+	err := svc.SetDesiredKubeObject(secret, comp.GetName()+"-provider-conf-credentials",
+		runtime.KubeOptionProtects("namespace-conditions"),
+		runtime.KubeOptionProtects("cluster"),
+		runtime.KubeOptionProtects(comp.GetName()+"-netpol"))
+	if err != nil {
+		svc.AddResult(runtime.NewWarningResult(fmt.Sprintf("cannot set credential secret for provider-sql: %s", err)))
+		svc.Log.Error(err, "cannot set credential secret for provider-sql")
+	}
+
+	tls := "preferred"
+	if comp.Spec.Parameters.TLS.TLSEnabled {
+		// We need to skip-verify here as with the postgresql.
+		// provider-sql does not support custom CA certs.
+		tls = "skip-verify"
+	}
+
+	config := &my1alpha1.ProviderConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: comp.GetName(),
+		},
+		Spec: my1alpha1.ProviderConfigSpec{
+			TLS: &tls,
+			Credentials: my1alpha1.ProviderCredentials{
+				Source: "MySQLConnectionSecret",
+				ConnectionSecretRef: &xpv1.SecretReference{
+					Name:      "provider-conf-credentials",
+					Namespace: comp.GetInstanceNamespace(),
+				},
+			},
+		},
+	}
+
+	err = svc.SetDesiredKubeObject(config, comp.GetName()+"-providerconfig")
+	if err != nil {
+		svc.AddResult(runtime.NewWarningResult(fmt.Sprintf("cannot apply the provider config for provider sql: %s", err)))
+		svc.Log.Error(err, "cannot apply the provider config for provider sql")
+	}
+}
+
+// We check if the database is already specified.
+// If not it will be added.
+// This should handle cases where there are mutliple users pointing to the same
+// database, and one is deleted, that the database is not dropped.
```

Comment:
how would You like to delete user or database? Fields in Claim stays forever and we do not have a switch like enabled/prune or anything like that
```
type VSHNAccess struct {
	// User specifies the username. If all other fields are left empty
	// then a new database with the same name and all permissions will be created.
	// +kubebuilder:validation:Required
	User *string `json:"user"`

	// Database is the name of the database to create, defaults to user.
	Database *string `json:"database,omitempty"`

	// Privileges specifies the privileges to grant the user. Please check
	// the database's docs for available privileges.
	Privileges []string `json:"privileges,omitempty"`

	// WriteConnectionSecretToReference specifies the namespace and name of a
	// Secret to which any connection details for this user should
	// be written.
	// If not specified, a secret with the name $claimname-$username will be
	// created in the namespace where the claim is located.
	WriteConnectionSecretToReference *xpv1.SecretReference `json:"writeConnectionSecretToRef,omitempty"`
}
```

Reply:
ok, I saw second MR (docs) and now I get how deletion works :D